Avoid task caching for static file volume and static block volume provisioning

chethanv28 opened the pull request:
<!-- Thanks for sending a pull request! -->

**What this PR does / why we need it**:
This PR is extending the current fix in master branch to remove taskInfo details from cache for static block volumes, to be used for static file volumes as well
This PR is making sure taskInfo details cache stores PV names only for dynamically provisioned volumes.
We do not need the cache for static volumes since they are not provisioned using external provisioner.
This is handled using backdiskId or backingfileId details in create spec.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
Recreated volumes statically with the same name
```
root@k8s-master:~# kubectl logs vsphere-csi-controller-79d8cf849-fjpss -c vsphere-syncer -n kube-system | grep "CreateVolume: Volume created successfully"
2020-12-15T03:30:38.161Z	INFO	volume/manager.go:308	CreateVolume: Volume created successfully. VolumeName: "static-pv-name", opId: "98cde46d", volumeID: "4a01a315-7d2f-4eae-bd73-a5edb51fa659"	{"TraceId": "0df6f951-fe48-48a5-8f1e-384d862daa5f"}
2020-12-15T03:32:37.110Z	INFO	volume/manager.go:308	CreateVolume: Volume created successfully. VolumeName: "static-pv-name", opId: "98cde48d", volumeID: "4d5b99f7-cc74-4bdc-86b7-1de2990aff0c"	{"TraceId": "35e1826d-166b-4e93-a5b7-165a4aa8ff36"}
```
**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Store taskInfo details in cache only for dynamically provisioned volumes
```

divyenpatel commented:
@chethanv28 Can you add a release-note and correct commit message on this PR?

divyenpatel commented on pkg/common/cns-lib/volume/manager.go:
```diff
@@ -279,10 +279,21 @@ func (m *defaultManager) CreateVolume(ctx context.Context, spec *cnstypes.CnsVol
+		isStaticallyProvisionedFileVolume = true
+	}
+
+	if isStaticallyProvisionedBlockVolume || isStaticallyProvisionedFileVolume {
```
Can you correct this condition as below?

`if !isStaticallyProvisionedBlockVolume && !isStaticallyProvisionedFileVolume {`

divyenpatel replied:
if that is the case then why do we need to add this check - `if isStaticallyProvisionedBlockVolume || isStaticallyProvisionedFileVolume {`?

We do not want to retrieve tasks only when the volume is statically provisioned.

chethanv28 replied:
Extending the static provisioning fix to file volumes

divyenpatel replied:
ok got it. here we are deleting a task. but can we not place the task in the map, if we determine volume being created is static volume.

chethanv28 replied:
That's reasonable too! Will move this check above, to avoid storing the tasks in the first place

chethanv28 replied:
Done! Verified with the latest commit:


```
kubectl logs vsphere-csi-controller-6b8fbfb48c-5c7r7 -c vsphere-syncer -n kube-system | grep "CreateVolume: Volume created successfully"
2020-12-16T19:59:32.430Z	INFO	volume/manager.go:301	CreateVolume: Volume created successfully. VolumeName: "static-pv-name", opId: "98ce3a03", volumeID: "54f1aade-bada-4d93-b0a6-fe6cecc765e9"	{"TraceId": "449e99a2-35e2-45da-9223-93cb60221b54"}
2020-12-16T20:02:26.469Z	INFO	volume/manager.go:301	CreateVolume: Volume created successfully. VolumeName: "static-pv-name", opId: "98ce3a2e", volumeID: "556798ca-5ed2-4da8-8ff8-ca428906bed2"	{"TraceId": "accf79a0-ca02-436d-9d3d-bd9cefeab392"}
```

divyenpatel commented on pkg/common/cns-lib/volume/manager.go:
```diff
@@ -279,10 +279,21 @@ func (m *defaultManager) CreateVolume(ctx context.Context, spec *cnstypes.CnsVol
 	}
 
-	blockBackingDetails, ok := spec.BackingObjectDetails.(*cnstypes.CnsBlockBackingDetails)
+	blockBackingDetails, isBlockBackingDetails := spec.BackingObjectDetails.(*cnstypes.CnsBlockBackingDetails)
```
As discussed Can you move the creation of blockBackingDetails and  fileBackingDetails objects inside the if condition after we identify block or file using `spec.VolumeType == string(cnstypes.CnsVolumeTypeBlock)` and `spec.VolumeType == string(cnstypes.CnsVolumeTypeFile)`.

chethanv28 reviewed (commented):
Addressed the recent review comments

chethanv28 commented:
jtest wcp

svcbot-qecnsdp commented:
Started WCP block pipeline...

chethanv28 commented:
jtest block-vanilla

svcbot-qecnsdp commented:
Started Vanilla block pipeline...

svcbot-qecnsdp commented:
``` 
WCP build status: FAILURE 
Stage before exit: null 

```

chethanv28 reviewed (commented):
Addressed review comments in latest commit. Rebase done

chethanv28 commented:
jtest wcp

svcbot-qecnsdp commented:
Started WCP block pipeline...

chethanv28 commented:
@divyenpatel I have addressed all the review comments. Can you help with review and approvals.

chethanv28 commented:
/assign @lipingxue Can you help review this PR

svcbot-qecnsdp commented:
``` 
WCP build status: FAILURE 
Stage before exit: null 
Jenkins E2E Test Results: 
Ran 31 of 155 Specs in 13193.589 seconds
FAIL! -- 21 Passed | 10 Failed | 0 Pending | 124 Skipped
--- FAIL: TestE2E (13193.64s)
FAIL

Ginkgo ran 1 suite in 4h7m5.483081603s
Test Suite Failed
make: Leaving directory `/home/worker/workspace/github-csi-wcp-CICD@2/Results/372/vsphere-csi-driver`

```

chethanv28 commented:
jtest wcp

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
``` 
WCP build status: FAILURE 
Stage before exit: null 
Jenkins E2E Test Results: 
Ran 31 of 155 Specs in 4590.439 seconds
FAIL! -- 21 Passed | 10 Failed | 0 Pending | 124 Skipped
--- FAIL: TestE2E (4590.49s)
FAIL

Ginkgo ran 1 suite in 1h18m41.352405381s
Test Suite Failed
make: Leaving directory `/home/worker/workspace/github-csi-wcp-CICD/Results/374/vsphere-csi-driver`

```

lipingxue commented:
> /assign @lipingxue
> Can you help review this PR

Change looks good to me.

chethanv28 commented:
> ```
> WCP build status: FAILURE 
> Stage before exit: null 
> Jenkins E2E Test Results: 
> Ran 31 of 155 Specs in 4590.439 seconds
> FAIL! -- 21 Passed | 10 Failed | 0 Pending | 124 Skipped
> --- FAIL: TestE2E (4590.49s)
> FAIL
> 
> Ginkgo ran 1 suite in 1h18m41.352405381s
> Test Suite Failed
> make: Leaving directory `/home/worker/workspace/github-csi-wcp-CICD/Results/374/vsphere-csi-driver`
> ```

All 10 failures are because of ENV variable not set! This looks clearly to be e2e test issue.

```
ENV NONSHARED_VSPHERE_DATASTORE_URL is not set
```

divyenpatel commented:
@chethanv28 I see we have 10 failed tests. Can you verify if failure is not due to this change?

divyenpatel commented:
Now you have handled the case for both block and file, can you change the PR title?

chethanv28 commented:
> @chethanv28 I see we have 10 failed tests. Can you verify if failure is not due to this change?
@divyenpatel 
Fails because of incorrect e2e environment in all 10 negative test cases :
```
12:13:23    ENV NONSHARED_VSPHERE_DATASTORE_URL is not set
12:13:23    Expected
12:13:23        <string>: 
12:13:23    not to be empty
``` 


Sample failures:
```
12:15:04  [Fail] Basic Static Provisioning [BeforeEach] [csi-supervisor] Verify static provisioning workflow - when DuplicateFCD is used 
12:15:04  /home/worker/workspace/github-csi-wcp-CICD/Results/374/vsphere-csi-driver/tests/e2e/e2e_common.go:149
12:15:04  
12:15:04  [Fail] Basic Static Provisioning [BeforeEach] [csi-supervisor] Verify static provisioning workflow SVC - On non shared datastore 
12:15:04  /home/worker/workspace/github-csi-wcp-CICD/Results/374/vsphere-csi-driver/tests/e2e/e2e_common.go:149
```

Deployed custom image built from this PR:

root@422c53ed536bf1dacce0d0c5e9b8b606 [ ~ ]# kubectl describe deployment vsphere-csi-controller -n vmware-system-csi | grep chethan
    Image:      vcsidev/vsphere-csi:chethan
    Image:      vcsidev/syncer:chethan

Ran these 2 tests manually:

```
$ export GINKGO_FOCUS="Verify static provisioning workflow SVC - On non shared datastore"
chethanv-a01:vsphere-csi-driver chethanv$ make test-e2e
hack/run-e2e-test.sh
go: found github.com/onsi/ginkgo/ginkgo in github.com/onsi/ginkgo v1.14.2
go: golang.org/x/sys upgrade => v0.0.0-20201221093633-bc327ba9c2f0
go: github.com/nxadm/tail upgrade => v1.4.5
go: github.com/fsnotify/fsnotify upgrade => v1.4.9
Dec 22 17:26:24.728: INFO: The --provider flag is not set. Continuing as if --provider=skeleton had been used.
W1222 17:26:24.728521   12082 test_context.go:414] Unable to find in-cluster config, using default host : http://127.0.0.1:8080
I1222 17:26:24.728645   12082 test_context.go:423] Tolerating taints "" when considering if nodes are ready
Running Suite: CNS CSI Driver End-to-End Tests
==============================================
Random Seed: 1608686776 - Will randomize all specs
Will run 1 of 155 specs

SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
Basic Static Provisioning 
  [csi-supervisor] Verify static provisioning workflow SVC - On non shared datastore
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/tests/e2e/csi_static_provisioning_basic.go:1277
[BeforeEach] Basic Static Provisioning
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/vendor/k8s.io/kubernetes/test/e2e/framework/framework.go:178
STEP: Creating a kubernetes client
Dec 22 17:26:24.750: INFO: >>> kubeConfig: /Users/chethanv/.kube/config
STEP: Building a namespace api object, basename e2e-csistaticprovision
Dec 22 17:26:25.175: INFO: Found PodSecurityPolicies; assuming PodSecurityPolicy is enabled.
Dec 22 17:26:25.377: INFO: Found ClusterRoles; assuming RBAC is enabled.
STEP: Binding the e2e-test-privileged-psp PodSecurityPolicy to the default service account in e2e-csistaticprovision-8081
STEP: Waiting for a default service account to be provisioned in namespace
[BeforeEach] Basic Static Provisioning
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/tests/e2e/csi_static_provisioning_basic.go:77
[It] [csi-supervisor] Verify static provisioning workflow SVC - On non shared datastore
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/tests/e2e/csi_static_provisioning_basic.go:1277
{"level":"info","time":"2020-12-22T17:26:26.869263-08:00","caller":"e2e/csi_static_provisioning_basic.go:1287","msg":"pvc name :cns-pvc-1608686786"}
STEP: Get the Policy ID and assign policy to storage class
Dec 22 17:26:27.222: INFO: storage policy id: e0a3843f-c12b-4023-ba8c-55387381cf47 for storage policy name is: nonshared
{"level":"info","time":"2020-12-22T17:26:27.222486-08:00","caller":"e2e/csi_static_provisioning_basic.go:1298","msg":"non shared datastore , profileID : e0a3843f-c12b-4023-ba8c-55387381cf47"}
STEP: Creating StorageClass nonshared with scParameters: map[storagePolicyID:e0a3843f-c12b-4023-ba8c-55387381cf47] and allowedTopologies: [] and ReclaimPolicy:  and allowVolumeExpansion: false
STEP: Creating Resource quota
{"level":"info","time":"2020-12-22T17:26:27.411457-08:00","caller":"e2e/csi_static_provisioning_basic.go:1308","msg":"storageclass Name :nonshared"}
STEP: Deleted Resource quota: cns-test-quota
STEP: Create Resource quota: &ResourceQuota{ObjectMeta:{cns-test-quota  e2e-test-namespace /api/v1/namespaces/e2e-test-namespace/resourcequotas/cns-test-quota 2677a03a-08df-44cb-b26a-452b89ca8d20 30862340 0 2020-12-22 17:26:27 -0800 PST <nil> <nil> map[] map[] [] []  [{e2e.test Update v1 2020-12-22 17:26:27 -0800 PST FieldsV1 FieldsV1{Raw:*[123 34 102 58 115 112 101 99 34 58 123 34 102 58 104 97 114 100 34 58 123 34 46 34 58 123 125 44 34 102 58 110 111 110 115 104 97 114 101 100 46 115 116 111 114 97 103 101 99 108 97 115 115 46 115 116 111 114 97 103 101 46 107 56 115 46 105 111 47 114 101 113 117 101 115 116 115 46 115 116 111 114 97 103 101 34 58 123 125 125 125 125],}}]},Spec:ResourceQuotaSpec{Hard:ResourceList{nonshared.storageclass.storage.k8s.io/requests.storage: {{214748364800 0} {<nil>}  BinarySI},},Scopes:[],ScopeSelector:nil,},Status:ResourceQuotaStatus{Hard:ResourceList{},Used:ResourceList{},},}
STEP: Waiting for 15 seconds to allow resourceQuota to be claimed
STEP: Create FCD
STEP: Sleeping for 90 seconds to allow newly created FCD:c4cd390d-d6f2-4701-b1cf-dcc62a27575e to sync with pandora
STEP: Create CNS register volume with above created FCD
{"level":"info","time":"2020-12-22T17:28:13.618583-08:00","caller":"e2e/util.go:1336","msg":"get CNSRegisterVolume spec"}
{"level":"info","time":"2020-12-22T17:28:13.950991-08:00","caller":"e2e/util.go:1367","msg":"Create CNSRegisterVolume"}
STEP: Verify the error message, when non-shared data store is used to created CRD
{"level":"info","time":"2020-12-22T17:28:34.089627-08:00","caller":"e2e/csi_static_provisioning_basic.go:1332","msg":"CNS register volume name : cnsregvol-vrzdn"}
STEP: Delete Resource quota
STEP: Deleted Resource quota: cns-test-quota
{"level":"info","time":"2020-12-22T17:28:34.599682-08:00","caller":"e2e/csi_static_provisioning_basic.go:1311","msg":"Delete storage class"}
[AfterEach] Basic Static Provisioning
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/vendor/k8s.io/kubernetes/test/e2e/framework/framework.go:179
Dec 22 17:28:34.708: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
STEP: Destroying namespace "e2e-csistaticprovision-8081" for this suite.
[AfterEach] Basic Static Provisioning
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/tests/e2e/csi_static_provisioning_basic.go:127
STEP: Performing test cleanup

• [SLOW TEST:130.138 seconds]
Basic Static Provisioning
/Users/chethanv/Downloads/csi-external/vsphere-csi-driver/tests/e2e/csi_static_provisioning_basic.go:53
  [csi-supervisor] Verify static provisioning workflow SVC - On non shared datastore
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/tests/e2e/csi_static_provisioning_basic.go:1277
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
Ran 1 of 155 Specs in 130.139 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 154 Skipped
PASS

Ginkgo ran 1 suite in 2m18.778148045s
Test Suite Passed
chethanv-a01:vsphere-csi-driver chethanv$ export GINKGO_FOCUS="Verify static provisioning workflow - when DuplicateFCD is used"
chethanv-a01:vsphere-csi-driver chethanv$ make test-e2e
hack/run-e2e-test.sh
go: found github.com/onsi/ginkgo/ginkgo in github.com/onsi/ginkgo v1.14.2
go: golang.org/x/sys upgrade => v0.0.0-20201221093633-bc327ba9c2f0
go: github.com/fsnotify/fsnotify upgrade => v1.4.9
go: github.com/nxadm/tail upgrade => v1.4.5
Dec 22 17:30:30.108: INFO: The --provider flag is not set. Continuing as if --provider=skeleton had been used.
W1222 17:30:30.107880   12250 test_context.go:414] Unable to find in-cluster config, using default host : http://127.0.0.1:8080
I1222 17:30:30.107997   12250 test_context.go:423] Tolerating taints "" when considering if nodes are ready
Running Suite: CNS CSI Driver End-to-End Tests
==============================================
Random Seed: 1608687022 - Will randomize all specs
Will run 1 of 155 specs

SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
Basic Static Provisioning 
  [csi-supervisor] Verify static provisioning workflow - when DuplicateFCD is used
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/tests/e2e/csi_static_provisioning_basic.go:895
[BeforeEach] Basic Static Provisioning
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/vendor/k8s.io/kubernetes/test/e2e/framework/framework.go:178
STEP: Creating a kubernetes client
Dec 22 17:30:30.126: INFO: >>> kubeConfig: /Users/chethanv/.kube/config
STEP: Building a namespace api object, basename e2e-csistaticprovision
Dec 22 17:30:30.663: INFO: Found PodSecurityPolicies; assuming PodSecurityPolicy is enabled.
Dec 22 17:30:30.832: INFO: Found ClusterRoles; assuming RBAC is enabled.
STEP: Binding the e2e-test-privileged-psp PodSecurityPolicy to the default service account in e2e-csistaticprovision-8081
STEP: Waiting for a default service account to be provisioned in namespace
[BeforeEach] Basic Static Provisioning
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/tests/e2e/csi_static_provisioning_basic.go:77
[It] [csi-supervisor] Verify static provisioning workflow - when DuplicateFCD is used
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/tests/e2e/csi_static_provisioning_basic.go:895
{"level":"info","time":"2020-12-22T17:30:32.263898-08:00","caller":"e2e/csi_static_provisioning_basic.go:905","msg":"pvc name :cns-pvc-1608687032"}
STEP: Get storage Policy
STEP: storagePolicyName: shared-ds-policy
Dec 22 17:30:32.607: INFO: storage policy id: 816802f7-8bcf-4a00-8e36-5bfdd1d33ff9 for storage policy name is: shared-ds-policy
{"level":"info","time":"2020-12-22T17:30:32.607402-08:00","caller":"e2e/csi_static_provisioning_basic.go:177","msg":"Profile ID :816802f7-8bcf-4a00-8e36-5bfdd1d33ff9"}
STEP: Creating StorageClass shared-ds-policy with scParameters: map[storagePolicyID:816802f7-8bcf-4a00-8e36-5bfdd1d33ff9] and allowedTopologies: [] and ReclaimPolicy:  and allowVolumeExpansion: false
{"level":"info","time":"2020-12-22T17:30:32.739853-08:00","caller":"e2e/csi_static_provisioning_basic.go:187","msg":"storageclass name :shared-ds-policy"}
STEP: create resource quota
{"level":"info","time":"2020-12-22T17:30:32.808272-08:00","caller":"e2e/csi_static_provisioning_basic.go:190","msg":"storageclass name :shared-ds-policy"}
STEP: Create Resource quota: &ResourceQuota{ObjectMeta:{cns-test-quota  e2e-test-namespace /api/v1/namespaces/e2e-test-namespace/resourcequotas/cns-test-quota e1ba95b2-4a41-4b1e-ada7-424404126a61 30864467 0 2020-12-22 17:30:32 -0800 PST <nil> <nil> map[] map[] [] []  [{e2e.test Update v1 2020-12-22 17:30:32 -0800 PST FieldsV1 FieldsV1{Raw:*[123 34 102 58 115 112 101 99 34 58 123 34 102 58 104 97 114 100 34 58 123 34 46 34 58 123 125 44 34 102 58 115 104 97 114 101 100 45 100 115 45 112 111 108 105 99 121 46 115 116 111 114 97 103 101 99 108 97 115 115 46 115 116 111 114 97 103 101 46 107 56 115 46 105 111 47 114 101 113 117 101 115 116 115 46 115 116 111 114 97 103 101 34 58 123 125 125 125 125],}}]},Spec:ResourceQuotaSpec{Hard:ResourceList{shared-ds-policy.storageclass.storage.k8s.io/requests.storage: {{214748364800 0} {<nil>}  BinarySI},},Scopes:[],ScopeSelector:nil,},Status:ResourceQuotaStatus{Hard:ResourceList{},Used:ResourceList{},},}
STEP: Waiting for 15 seconds to allow resourceQuota to be claimed
STEP: Creating Resource quota
STEP: Deleted Resource quota: cns-test-quota
STEP: Create Resource quota: &ResourceQuota{ObjectMeta:{cns-test-quota  e2e-test-namespace /api/v1/namespaces/e2e-test-namespace/resourcequotas/cns-test-quota a6fc7f54-6ff8-4749-a8c7-ac6ac8d82d03 30864597 0 2020-12-22 17:30:48 -0800 PST <nil> <nil> map[] map[] [] []  [{e2e.test Update v1 2020-12-22 17:30:48 -0800 PST FieldsV1 FieldsV1{Raw:*[123 34 102 58 115 112 101 99 34 58 123 34 102 58 104 97 114 100 34 58 123 34 46 34 58 123 125 44 34 102 58 115 104 97 114 101 100 45 100 115 45 112 111 108 105 99 121 46 115 116 111 114 97 103 101 99 108 97 115 115 46 115 116 111 114 97 103 101 46 107 56 115 46 105 111 47 114 101 113 117 101 115 116 115 46 115 116 111 114 97 103 101 34 58 123 125 125 125 125],}}]},Spec:ResourceQuotaSpec{Hard:ResourceList{shared-ds-policy.storageclass.storage.k8s.io/requests.storage: {{214748364800 0} {<nil>}  BinarySI},},Scopes:[],ScopeSelector:nil,},Status:ResourceQuotaStatus{Hard:ResourceList{},Used:ResourceList{},},}
STEP: Waiting for 15 seconds to allow resourceQuota to be claimed
STEP: Create FCD
STEP: Sleeping for 90 seconds to allow newly created FCD: to sync with pandora
STEP: Create CNS register volume with above created FCD 
{"level":"info","time":"2020-12-22T17:34:41.168012-08:00","caller":"e2e/util.go:1336","msg":"get CNSRegisterVolume spec"}
{"level":"info","time":"2020-12-22T17:34:41.575598-08:00","caller":"e2e/util.go:1367","msg":"Create CNSRegisterVolume"}
Dec 22 17:34:41.704: INFO: Waiting up to 3m0s for CnsRegisterVolume &{{ } {cnsregvol-5kzdj cnsregvol- e2e-test-namespace /apis/cns.vmware.com/v1alpha1/namespaces/e2e-test-namespace/cnsregistervolumes/cnsregvol-5kzdj 456c6dc9-34f7-41e2-b859-d0be80ed333d 30866589 %!s(int64=1) 2020-12-22 17:34:41 -0800 PST <nil> %!s(*int64=<nil>) map[] map[] [] []  [{e2e.test Update cns.vmware.com/v1alpha1 2020-12-22 17:34:41 -0800 PST FieldsV1 &FieldsV1{Raw:*[123 34 102 58 109 101 116 97 100 97 116 97 34 58 123 34 102 58 103 101 110 101 114 97 116 101 78 97 109 101 34 58 123 125 125 44 34 102 58 115 112 101 99 34 58 123 34 46 34 58 123 125 44 34 102 58 97 99 99 101 115 115 77 111 100 101 34 58 123 125 44 34 102 58 112 118 99 78 97 109 101 34 58 123 125 44 34 102 58 118 111 108 117 109 101 73 68 34 58 123 125 125 44 34 102 58 115 116 97 116 117 115 34 58 123 34 46 34 58 123 125 44 34 102 58 114 101 103 105 115 116 101 114 101 100 34 58 123 125 125 125],}}]} {cns-pvc-1608687032 c7df96b3-d7e2-46e6-ab46-7b508f0e05ca ReadWriteOnce } {%!s(bool=false) }} to get created
Dec 22 17:34:42.117: INFO: cnsRegisterVolume cnsregvol-5kzdj found and Registered status is  =%!s(bool=false) (413.010434ms)
Dec 22 17:34:44.436: INFO: cnsRegisterVolume cnsregvol-5kzdj found and Registered status is  =%!s(bool=false) (2.732213884s)
Dec 22 17:34:46.890: INFO: cnsRegisterVolume cnsregvol-5kzdj found and Registered status is  =%!s(bool=false) (5.186452749s)
Dec 22 17:34:49.413: INFO: cnsRegisterVolume cnsregvol-5kzdj found and Registered status is  =%!s(bool=false) (7.708648147s)
Dec 22 17:34:52.605: INFO: cnsRegisterVolume cnsregvol-5kzdj found and Registered status is  =%!s(bool=false) (10.901381698s)
STEP: verify created PV, PVC
{"level":"info","time":"2020-12-22T17:34:55.110331-08:00","caller":"e2e/csi_static_provisioning_basic.go:933","msg":"CNS register volume name : cnsregvol-5kzdj"}
STEP: Create CnsregisteVolume with already used FCD
{"level":"info","time":"2020-12-22T17:34:55.280127-08:00","caller":"e2e/util.go:1336","msg":"get CNSRegisterVolume spec"}
{"level":"info","time":"2020-12-22T17:34:55.719158-08:00","caller":"e2e/util.go:1367","msg":"Create CNSRegisterVolume"}
Dec 22 17:34:55.870: INFO: Waiting up to 3m0s for CnsRegisterVolume &{{ } {cnsregvol-t7rql cnsregvol- e2e-test-namespace /apis/cns.vmware.com/v1alpha1/namespaces/e2e-test-namespace/cnsregistervolumes/cnsregvol-t7rql 6e813243-8c2a-4596-8389-a282ae000bd5 30866725 %!s(int64=1) 2020-12-22 17:34:55 -0800 PST <nil> %!s(*int64=<nil>) map[] map[] [] []  [{e2e.test Update cns.vmware.com/v1alpha1 2020-12-22 17:34:55 -0800 PST FieldsV1 &FieldsV1{Raw:*[123 34 102 58 109 101 116 97 100 97 116 97 34 58 123 34 102 58 103 101 110 101 114 97 116 101 78 97 109 101 34 58 123 125 125 44 34 102 58 115 112 101 99 34 58 123 34 46 34 58 123 125 44 34 102 58 97 99 99 101 115 115 77 111 100 101 34 58 123 125 44 34 102 58 112 118 99 78 97 109 101 34 58 123 125 44 34 102 58 118 111 108 117 109 101 73 68 34 58 123 125 125 44 34 102 58 115 116 97 116 117 115 34 58 123 34 46 34 58 123 125 44 34 102 58 114 101 103 105 115 116 101 114 101 100 34 58 123 125 125 125],}}]} {cns-pvc-1608687032duplicatefcd c7df96b3-d7e2-46e6-ab46-7b508f0e05ca ReadWriteOnce } {%!s(bool=false) }} to get created
Dec 22 17:34:56.260: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (389.533805ms)
Dec 22 17:34:58.721: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (2.851399283s)
Dec 22 17:35:01.301: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (5.430493333s)
Dec 22 17:35:03.653: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (7.783096294s)
Dec 22 17:35:06.031: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (10.160735717s)
Dec 22 17:35:08.374: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (12.503822454s)
Dec 22 17:35:10.737: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (14.866264211s)
Dec 22 17:35:13.095: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (17.22465442s)
Dec 22 17:35:15.522: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (19.651967565s)
Dec 22 17:35:18.027: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (22.1567037s)
Dec 22 17:35:20.681: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (24.810691043s)
Dec 22 17:35:23.026: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (27.155256539s)
Dec 22 17:35:25.399: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (29.527915425s)
Dec 22 17:35:27.707: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (31.835859535s)
Dec 22 17:35:30.106: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (34.235133265s)
Dec 22 17:35:32.433: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (36.562276648s)
Dec 22 17:35:34.898: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (39.02746109s)
Dec 22 17:35:37.231: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (41.360089029s)
Dec 22 17:35:39.571: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (43.699783913s)
Dec 22 17:35:42.087: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (46.216362066s)
Dec 22 17:35:44.427: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (48.556197015s)
Dec 22 17:35:46.810: INFO: cnsRegisterVolume cnsregvol-t7rql found and Registered status is  =%!s(bool=false) (50.939279893s)
SSSSSSSSSSSSSSSSSSSSSSSSSSSS
Ran 1 of 155 Specs in 130.139 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 154 Skipped
PASS

Ginkgo ran 1 suite in 2m18.778148045s
Test Suite Passed
```

divyenpatel commented:
/lgtm

divyenpatel commented:
/hold

divyenpatel commented on pkg/common/cns-lib/volume/manager.go:
```diff
@@ -206,13 +206,31 @@ func (m *defaultManager) CreateVolume(ctx context.Context, spec *cnstypes.CnsVol
+		if !isStaticallyProvisionedBlockVolume && !isStaticallyProvisionedFileVolume {
+			var taskDetails createVolumeTaskDetails
+			// Store the task details and task object expiration time in volumeTaskMap
+			taskDetails.Lock()
```
I observed this late but can we remove lock on locally defined variable? Lock and unlock on local variable is not useful. We can directly save the task in the Map.

chethanv28 replied:
Ok, makes sense

chethanv28 commented:
Tested again with new images after addressing latest comments:

vanilla cluster:

```
root@k8s-master-825:~# kubectl get pvc -A
NAMESPACE   NAME                        STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS               AGE
default     example-vanilla-block-pvc   Bound    pvc-0a5dbe7d-72f8-4825-ad63-c22c4ff7aac9   1Mi        RWO            example-vanilla-block-sc   4m32s
default     example-vanilla-file-pvc    Bound    pvc-e057e5c0-00dc-43b0-86c9-02d9d6a48856   1Mi        RWX            example-vanilla-file-sc    77s
root@k8s-master-825:~# kubectl get pods
NAME                        READY   STATUS    RESTARTS   AGE
example-vanilla-block-pod   1/1     Running   0          2m21s
root@k8s-master-825:~# kubectl describe deployment vsphere-csi-controller -n kube-system | grep chethan
    Image:      vcsidev/vsphere-csi:chethan-1
    Image:      vcsidev/syncer:chethan-1
```
cc: @divyenpatel

chethanv28 reviewed (commented):
Reduced provisioner timeout to simulate orphan volume creation scenario.
Verified that the changes in the PR work as expected to avoid new create volume calls and consider tasks in cache:


```
2020-12-23T03:06:19.493Z	DEBUG	volume/manager.go:180	Update VSphereUser from Administrator@vsphere.local to VSPHERE.LOCAL\Administrator	{"TraceId": "c3d713c4-e55f-410a-ab48-49f729374c48"}
2020-12-23T03:06:19.499Z	INFO	volume/manager.go:194	CreateVolume task still pending for VolumeName: "pvc-e057e5c0-00dc-43b0-86c9-02d9d6a48856", with taskInfo: Task:task-4226	{"TraceId": "c3d713c4-e55f-410a-ab48-49f729374c48"}
2020-12-23T03:06:19.521Z	INFO	volume/manager.go:239	CreateVolume: VolumeName: "pvc-e057e5c0-00dc-43b0-86c9-02d9d6a48856", opId: "53df2f41"	{"TraceId": "c3d713c4-e55f-410a-ab48-49f729374c48"}
2020-12-23T03:06:19.525Z	INFO	volume/manager.go:298	CreateVolume: Volume created successfully. VolumeName: "pvc-e057e5c0-00dc-43b0-86c9-02d9d6a48856", opId: "53df2f41", volumeID: "file:3b263dcb-4425-4c5a-b79b-8ed2c06722f7"	{"TraceId": "c3d713c4-e55f-410a-ab48-49f729374c48"}
2020-12-23T03:06:19.525Z	DEBUG	volume/manager.go:299	CreateVolume volumeId "file:3b263dcb-4425-4c5a-b79b-8ed2c06722f7" is placed on datastore "ds:///vmfs/volumes/vsan:5241290d7693972b-deddb79aced86eb5/"	{"TraceId": "c3d713c4-e55f-410a-ab48-49f729374c48"}
```

Supervisor Cluster static provisioning tests:
```
$ export GINKGO_FOCUS="Verify static provisioning workflow SVC - On non shared datastore"
chethanv-a01:vsphere-csi-driver chethanv$ make test-e2e
hack/run-e2e-test.sh
go: found github.com/onsi/ginkgo/ginkgo in github.com/onsi/ginkgo v1.14.2
go: golang.org/x/sys upgrade => v0.0.0-20201221093633-bc327ba9c2f0
go: github.com/nxadm/tail upgrade => v1.4.5
go: github.com/fsnotify/fsnotify upgrade => v1.4.9
Dec 22 19:11:40.283: INFO: The --provider flag is not set. Continuing as if --provider=skeleton had been used.
W1222 19:11:40.283035   25735 test_context.go:414] Unable to find in-cluster config, using default host : http://127.0.0.1:8080
I1222 19:11:40.283135   25735 test_context.go:423] Tolerating taints "" when considering if nodes are ready
Running Suite: CNS CSI Driver End-to-End Tests
==============================================
Random Seed: 1608693092 - Will randomize all specs
Will run 1 of 155 specs

SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
Basic Static Provisioning 
  [csi-supervisor] Verify static provisioning workflow SVC - On non shared datastore
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/tests/e2e/csi_static_provisioning_basic.go:1277
[BeforeEach] Basic Static Provisioning
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/vendor/k8s.io/kubernetes/test/e2e/framework/framework.go:178
STEP: Creating a kubernetes client
Dec 22 19:11:40.299: INFO: >>> kubeConfig: /Users/chethanv/.kube/config
STEP: Building a namespace api object, basename e2e-csistaticprovision
Dec 22 19:11:40.761: INFO: Found PodSecurityPolicies; assuming PodSecurityPolicy is enabled.
Dec 22 19:11:41.045: INFO: Found ClusterRoles; assuming RBAC is enabled.
STEP: Binding the e2e-test-privileged-psp PodSecurityPolicy to the default service account in e2e-csistaticprovision-8081
STEP: Waiting for a default service account to be provisioned in namespace
[BeforeEach] Basic Static Provisioning
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/tests/e2e/csi_static_provisioning_basic.go:77
[It] [csi-supervisor] Verify static provisioning workflow SVC - On non shared datastore
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/tests/e2e/csi_static_provisioning_basic.go:1277
{"level":"info","time":"2020-12-22T19:11:42.391536-08:00","caller":"e2e/csi_static_provisioning_basic.go:1287","msg":"pvc name :cns-pvc-1608693102"}
STEP: Get the Policy ID and assign policy to storage class
Dec 22 19:11:42.736: INFO: storage policy id: e0a3843f-c12b-4023-ba8c-55387381cf47 for storage policy name is: nonshared
{"level":"info","time":"2020-12-22T19:11:42.736231-08:00","caller":"e2e/csi_static_provisioning_basic.go:1298","msg":"non shared datastore , profileID : e0a3843f-c12b-4023-ba8c-55387381cf47"}
STEP: Creating StorageClass nonshared with scParameters: map[storagePolicyID:e0a3843f-c12b-4023-ba8c-55387381cf47] and allowedTopologies: [] and ReclaimPolicy:  and allowVolumeExpansion: false
STEP: Creating Resource quota
{"level":"info","time":"2020-12-22T19:11:42.958837-08:00","caller":"e2e/csi_static_provisioning_basic.go:1308","msg":"storageclass Name :nonshared"}
STEP: Deleted Resource quota: cns-test-quota
STEP: Create Resource quota: &ResourceQuota{ObjectMeta:{cns-test-quota  e2e-test-namespace /api/v1/namespaces/e2e-test-namespace/resourcequotas/cns-test-quota ecfa8429-50ab-483e-9428-de741d483d5a 30917014 0 2020-12-22 19:11:43 -0800 PST <nil> <nil> map[] map[] [] []  [{e2e.test Update v1 2020-12-22 19:11:43 -0800 PST FieldsV1 FieldsV1{Raw:*[123 34 102 58 115 112 101 99 34 58 123 34 102 58 104 97 114 100 34 58 123 34 46 34 58 123 125 44 34 102 58 110 111 110 115 104 97 114 101 100 46 115 116 111 114 97 103 101 99 108 97 115 115 46 115 116 111 114 97 103 101 46 107 56 115 46 105 111 47 114 101 113 117 101 115 116 115 46 115 116 111 114 97 103 101 34 58 123 125 125 125 125],}}]},Spec:ResourceQuotaSpec{Hard:ResourceList{nonshared.storageclass.storage.k8s.io/requests.storage: {{214748364800 0} {<nil>}  BinarySI},},Scopes:[],ScopeSelector:nil,},Status:ResourceQuotaStatus{Hard:ResourceList{},Used:ResourceList{},},}
STEP: Waiting for 15 seconds to allow resourceQuota to be claimed
STEP: Create FCD
STEP: Sleeping for 90 seconds to allow newly created FCD:d1576f4b-6a61-49ed-aede-aa78f75d9b27 to sync with pandora
STEP: Create CNS register volume with above created FCD
{"level":"info","time":"2020-12-22T19:13:29.454548-08:00","caller":"e2e/util.go:1336","msg":"get CNSRegisterVolume spec"}
{"level":"info","time":"2020-12-22T19:13:29.906357-08:00","caller":"e2e/util.go:1367","msg":"Create CNSRegisterVolume"}
STEP: Verify the error message, when non-shared data store is used to created CRD
{"level":"info","time":"2020-12-22T19:13:50.055972-08:00","caller":"e2e/csi_static_provisioning_basic.go:1332","msg":"CNS register volume name : cnsregvol-8xz24"}
STEP: Delete Resource quota
STEP: Deleted Resource quota: cns-test-quota
{"level":"info","time":"2020-12-22T19:13:50.845048-08:00","caller":"e2e/csi_static_provisioning_basic.go:1311","msg":"Delete storage class"}
[AfterEach] Basic Static Provisioning
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/vendor/k8s.io/kubernetes/test/e2e/framework/framework.go:179
Dec 22 19:13:50.959: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
STEP: Destroying namespace "e2e-csistaticprovision-8081" for this suite.
[AfterEach] Basic Static Provisioning
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/tests/e2e/csi_static_provisioning_basic.go:127
STEP: Performing test cleanup

• [SLOW TEST:130.803 seconds]
Basic Static Provisioning
/Users/chethanv/Downloads/csi-external/vsphere-csi-driver/tests/e2e/csi_static_provisioning_basic.go:53
  [csi-supervisor] Verify static provisioning workflow SVC - On non shared datastore
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/tests/e2e/csi_static_provisioning_basic.go:1277
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
Ran 1 of 155 Specs in 130.804 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 154 Skipped
PASS


```

divyenpatel commented:
/hold cancel

divyenpatel commented:
/lgtm
